worker/runcworker/worker.go

```diff
@@ -114,32 +134,121 @@ func (w *runcworker) Exec(ctx context.Context, meta worker.Meta, root cache.Moun
 		return err
 	}
 
-	logrus.Debugf("> running %s %v", id, meta.Args)
+	forwardIO, err := newForwardIO(stdin, stdout, stderr)
+	if err != nil {
+		return errors.Wrap(err, "creating new forwarding IO")
+	}
+	defer forwardIO.Close()
+
+	pidFilePath := filepath.Join(w.root, "runc_pid_"+identity.NewID())
+	defer os.RemoveAll(pidFilePath)
+
+	logrus.Debugf("> creating %s %v", id, meta.Args)
+	if err := w.runc.Create(ctx, id, bundle, &runc.CreateOpts{
+		PidFile: pidFilePath,
+		IO:      forwardIO,
+	}); err != nil {
+		return err
+	}
+	forwardIO.release()
+
+	defer func() {
+		if err := w.runc.Delete(context.TODO(), id, &runc.DeleteOpts{}); err != nil {
+			logrus.Errorf("failed to delete %s: %+v", id, err)
+		}
+	}()
 
-	status, err := w.runc.Run(ctx, id, bundle, &runc.CreateOpts{
-		IO: &forwardIO{stdin: stdin, stdout: stdout, stderr: stderr},
-	})
-	logrus.Debugf("< completed %s %v %v", id, status, err)
-	if status != 0 {
-		select {
-		case <-ctx.Done():
-			// runc can't report context.Cancelled directly
-			return errors.Wrapf(ctx.Err(), "exit code %d", status)
-		default:
+	dt, err := ioutil.ReadFile(pidFilePath)
+	if err != nil {
+		return err
+	}
+	pid, err := strconv.Atoi(string(dt))
+	if err != nil {
+		return err
+	}
+
+	if iface != nil {
+		if err := iface.Set(pid); err != nil {
+			return errors.Wrap(err, "could not set the network")
 		}
-		return errors.Errorf("exit code %d", status)
+		defer iface.Remove()
 	}
 
-	return err
+	err = w.runc.Start(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	p, err := os.FindProcess(pid)
+	if err != nil {
+		return err
+	}
+
+	if ps, err := p.Wait(); err != nil {
```

Comment:
Issue in my example is causing the test failure. Contrary to `cmd.Wait()`, `process.Wait()` does not need to return an error for the exit code to be non-zero. So move the status code outside of this condition.